std/experimental/testing/io.d

```diff
@@ -0,0 +1,312 @@
+     */
+    void write(T...)(T args)
+    {
+        _tid.send(text(args));
+    }
+
+    /**
+     * Writes the args in a thread-safe manner and appends a newline.
+     */
+    void writeln(T...)(T args)
+    {
+        write(args, "\n");
+    }
+
+    /**
+     * Writes the args in a thread-safe manner in green (POSIX only).
+     * and appends a newline.
+     */
+    void writelnGreen(T...)(T args)
+    {
+        _tid.send(green(text(args) ~ "\n"));
+    }
+
+    /**
+     * Writes the args in a thread-safe manner in red (POSIX only)
+     * and appends a newline.
+     */
+    void writelnRed(T...)(T args)
+    {
+        _tid.send(red(text(args) ~ "\n"));
+    }
+
+    /**
+     * Writes the args in a thread-safe manner in red (POSIX only).
+     * and appends a newline.
+     */
+    void writeRed(T...)(T args)
+    {
+        _tid.send(red(text(args)));
+    }
+
+    /**
+     * Writes the args in a thread-safe manner in yellow (POSIX only).
+     * and appends a newline.
+     */
+    void writeYellow(T...)(T args)
+    {
+        _tid.send(yellow(text(args)));
+    }
+
+    /**
+     * Creates the singleton instance and waits until it's ready.
+     */
+    static void start()
+    {
+        WriterThread.get._tid.send(true, thisTid);
+        receiveOnly!bool; //wait for it to start
+    }
+
+    /**
+     * Waits for the writer thread to terminate.
+     */
+    void join()
+    {
+        _tid.send(thisTid); //tell it to join
+        receiveOnly!Tid(); //wait for it to join
+        _instance = null;
+        _instantiated = false;
+    }
+
+private:
+
+    this()
+    {
+        _tid = spawn(&threadWriter);
+        _escCodes = [ "red": "\033[31;1m",
+                      "green": "\033[32;1m",
+                      "yellow": "\033[33;1m",
+                      "cancel": "\033[0;;m" ];
+
+        version(Posix)
+        {
+            import core.sys.posix.unistd;
+            _useEscCodes = _forceEscCodes || isatty(stdout.fileno()) != 0;
+        }
+    }
+
+    /**
+     * Generate green coloured output on POSIX systems
+     */
+    string green(in string msg) @safe pure const
+    {
+        return escCode("green") ~ msg ~ escCode("cancel");
+    }
+
+    /**
+     * Generate red coloured output on POSIX systems
+     */
+    string red(in string msg) @safe pure const
+    {
+        return escCode("red") ~ msg ~ escCode("cancel");
+    }
+
+    /**
+     * Generate yellow coloured output on POSIX systems
+     */
+    string yellow(in string msg) @safe pure const
+    {
+        return escCode("yellow") ~ msg ~ escCode("cancel");
+    }
+
+    /**
+     * Send escape code to the console
+     */
+    string escCode(in string code) @safe pure const
+    {
+        return _useEscCodes ? _escCodes[code] : "";
+    }
+
+
+    Tid _tid;
+    string[string] _escCodes;
+    bool _useEscCodes;
+
+    static bool _instantiated; /// Thread local
+    __gshared WriterThread _instance;
+}
+
+private void threadWriter()
```

Comment:
I hadn't even connected those this to asynclogger! I wrote this code more than a year ago, it's been the same for most of its life (I think, git might prove me wrong).